File: src/app/services/policy-holder.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ProposalType, PolicyHolder, MailingAddressDTO, BankAccountDetails, VerificationStatus, PassportDetailsDTO, EiaDetailsDTO, LanguageDTO, PolicyHolderSearchRequest, PolicyHolderSearchResponse } from '../models';

@Injectable({
  providedIn: 'root'
})
export class PolicyHolderService {
  private baseUrl = 'http://localhost:8080/api';

  constructor(private http: HttpClient) {}

  getDropdownData(): Observable<any> {
    return this.http.get(`${this.baseUrl}/dropdown-data`);
  }

  savePolicyHolder(policyHolder: PolicyHolder): Observable<any> {
    return this.http.post(`${this.baseUrl}/policy-holder`, policyHolder);
  }

  validateGlobalLoadingFlag(): boolean {
    // Logic to return the value of the global loading flag
    return true; // Placeholder logic
  }

  fetchProposalTypes(): Observable<ProposalType[]> {
    return this.http.get<ProposalType[]>(`${this.baseUrl}/proposal-types`);
  }

  validateAgeProof(ageProof: string): Observable<any> {
    return this.http.post(`${this.baseUrl}/validate-age-proof`, { ageProof });
  }

  updateFormStatus(formStatus: string): Observable<any> {
    return this.http.post(`${this.baseUrl}/update-form-status`, { formStatus });
  }

  updatePolicyHolderDetails(policyHolder: PolicyHolder): Observable<any> {
    return this.http.put(`${this.baseUrl}/policy-holder`, policyHolder);
  }

  clearPolicyHolderDetails(): Observable<any> {
    return this.http.delete(`${this.baseUrl}/policy-holder`);
  }

  getValidPaCodes(): Observable<string[]> {
    return this.http.get<string[]>(`${this.baseUrl}/valid-pa-codes`);
  }

  validatePaCode(paCode: string): Observable<boolean> {
    return this.http.post<boolean>(`${this.baseUrl}/validate-pa-code`, { paCode });
  }

  updateTelephone2(telephone2: string): Observable<any> {
    return this.http.put(`${this.baseUrl}/update-telephone2`, { telephone2 });
  }

  validatePanIssuanceDate(panIssuanceDate: string): Observable<boolean> {
    return this.http.post<boolean>(`${this.baseUrl}/validate-pan-issuance-date`, { panIssuanceDate });
  }

  searchMailingAddress(partnerId: string): Observable<MailingAddressDTO> {
    return this.http.get<MailingAddressDTO>(`${this.baseUrl}/mailing-address/${partnerId}`);
  }

  savePanIssuanceDateState(isAvailable: boolean): void {
    // Logic to save the state of PAN issuance date availability
  }

  getBankAccountDetails(applicationNumber: string): Observable<BankAccountDetails> {
    return this.http.get<BankAccountDetails>(`${this.baseUrl}/bank-account-details/${applicationNumber}`);
  }

  getVerificationStatus(applicationNumber: string): Observable<VerificationStatus> {
    return this.http.get<VerificationStatus>(`${this.baseUrl}/verification-status/${applicationNumber}`);
  }

  validateAgeProofDetails(ageProofType: string, ageProofID: string, agentCode: string): Observable<any> {
    return this.http.post(`${this.baseUrl}/validate-age-proof-details`, { ageProofType, ageProofID, agentCode });
  }

  fetchPassportDetails(): Observable<PassportDetailsDTO> {
    return this.http.get<PassportDetailsDTO>(`${this.baseUrl}/passport-details`);
  }

  verifyVoterId(): Observable<any> {
    return this.http.get(`${this.baseUrl}/verify-voter-id`);
  }

  getDrivingLicenseDetails(): Observable<DrivingLicenseDetailsDTO> {
    return this.http.get<DrivingLicenseDetailsDTO>(`${this.baseUrl}/driving-license-details`);
  }

  createParameterList(): object {
    // Logic to create a new parameter list and add necessary parameters
    return {};
  }

  callFinancialDocumentForm(parameterList: object): void {
    this.http.post(`${this.baseUrl}/call-financial-document-form`, parameterList).subscribe();
  }

  getEiaDetails(applicationNumber: string): Observable<EiaDetailsDTO> {
    return this.http.get<EiaDetailsDTO>(`${this.baseUrl}/eia-details/${applicationNumber}`);
  }

  openEIAccount(): Observable<any> {
    return this.http.post(`${this.baseUrl}/open-ei-account`, {});
  }

  searchPolicyHolder(policyHolderSearchRequest: PolicyHolderSearchRequest): Observable<PolicyHolderSearchResponse> {
    return this.http.post<PolicyHolderSearchResponse>(`${this.baseUrl}/search-policy-holder`, policyHolderSearchRequest);
  }

  fetchLanguages(): Observable<LanguageDTO[]> {
    return this.http.get<LanguageDTO[]>(`${this.baseUrl}/languages`);
  }

  updateCommunicationLanguage(languageId: number): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/update-communication-language`, { languageId });
  }

  saveDropdownValue(value: string): void {
    // Logic to save the selected value of the dropdown list
  }

  getDropdownOptions(): string[] {
    // Logic to return the predefined options for the dropdown list
    return [];
  }

  getRelationshipType(): string {
    // Logic to return the relationship type as a string
    return '';
  }
}

File: src/main/java/com/azbj/fm2002/service/PolicyHolderService.java
package com.azbj.fm2002.service;

import com.azbj.fm2002.repository.PolicyHolderRepository;
import com.azbj.fm2002.dto.*;
import com.azbj.fm2002.util.AgeCalculationUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class PolicyHolderService {

    @Autowired
    private PolicyHolderRepository policyHolderRepository;

    public List<ProposalType> fetchProposalTypes() {
        return policyHolderRepository.findAllProposalTypes();
    }

    public String validateAgeProof(String ageProof) {
        try {
            return policyHolderRepository.findAgeUidByAgeProof(ageProof);
        } catch (NoDataFoundException e) {
            return "NA1";
        } catch (Exception e) {
            return "NA2";
        }
    }

    public String updateFormStatus(String formStatus) {
        int result = policyHolderRepository.updateFormStatus(formStatus);
        return result > 0 ? "Success" : "Failure";
    }

    public void updatePolicyHolderDetails(PolicyHolder policyHolder) {
        policyHolderRepository.save(policyHolder);
    }

    public void clearPolicyHolderDetails(PolicyHolder policyHolder) {
        policyHolderRepository.delete(policyHolder);
    }

    public boolean validateAccountNumber(String accountNumber) {
        return accountNumber != null && accountNumber.length() == 16;
    }

    public boolean validatePaCode(String paCode) {
        return policyHolderRepository.validatePaCode(paCode);
    }

    public boolean validatePanIssuanceDate(String panIssuanceDate) {
        return panIssuanceDate.compareTo(java.time.LocalDate.now().toString()) <= 0;
    }

    public MailingAddressDTO fetchMailingAddress(String partnerId) {
        return policyHolderRepository.findMailingAddress(partnerId);
    }

    public PaymentMethodValidationResponse validatePaymentMethod(PaymentMethodValidationRequest request) {
        return policyHolderRepository.validatePaymentMethod(request);
    }

    public BankAccountDetails getBankAccountDetails(String applicationNumber) {
        return policyHolderRepository.getBankAccountDetails(applicationNumber);
    }

    public VerificationStatus getVerificationStatus(String applicationNumber) {
        return policyHolderRepository.getVerificationStatus(applicationNumber);
    }

    public ValidationResponse validate(String ageProofType, String ageProofID, String agentCode) {
        ValidationResponse response = new ValidationResponse();
        if ("AC".equals(ageProofType) || "ACS".equals(ageProofType)) {
            if (ageProofID.length() == 12) {
                // Extract first name and last name from policy holder's name
                String firstName = "John"; // Placeholder
                String lastName = "Doe"; // Placeholder
                // Format date of birth
                String dob = "1990-01-01"; // Placeholder
                // Validate Aadhaar details
                String validatedDetails = "Validated"; // Placeholder
                policyHolderRepository.updateAgeProofDetails(ageProofID, validatedDetails);
                response.setMessage("Aadhaar validated");
            }
        } else if ("SYBM".equals(ageProofType) && !agentCode.startsWith("522")) {
            response.setMessage("Syndicate Bank BM Certificate is allowed only for Syndicate Bank");
        } else if ("Pan Card".equals(ageProofType) && ageProofID != null) {
            policyHolderRepository.updateAgeProofDetails(ageProofID, "Y");
            response.setMessage("PAN Card validated");
        }
        return response;
    }

    public PassportDetailsDTO fetchPassportDetails() {
        return policyHolderRepository.findPassportDetails();
    }

    public void verifyVoterId() {
        String applicationNumber = policyHolderRepository.getApplicationNumber("signCardNumber");
        String ageProof = policyHolderRepository.getAgeProof("insuredPersonId");
        if ("VI".equals(ageProof)) {
            String state = policyHolderRepository.getStateOfResidence("policyHolderId");
            // Call verification form with parameters
        } else {
            // Display message indicating Voter's ID details are not selected
        }
    }

    public DrivingLicenseDetailsDTO getDrivingLicenseDetails() {
        return policyHolderRepository.findDrivingLicenseDetails();
    }

    public void handleFinancialDocumentForm(Object parameterList) {
        // Process the parameter list and perform necessary actions
    }

    public EiaDetailsDTO getEiaDetails(String applicationNumber) {
        return policyHolderRepository.findEiaDetails(applicationNumber);
    }

    public void openEIAccount() {
        String applicationNo = "appNo"; // Placeholder
        PolicyHolderDetails details = policyHolderRepository.findPolicyHolderDetails(applicationNo);
        boolean exists = policyHolderRepository.checkExistingEIAccount(applicationNo);
        if (!exists) {
            policyHolderRepository.insertNewEIAccount(details);
        }
    }

    public void verifyAadhaar(String aadhaarNumber, String moduleName) {
        // Create a parameter list named 'Param1'
        // Add the Aadhaar number and module name ('BBU') to the parameter list
        // Increment the eKYC review count
        // Call the form 'AZBJ_ADHAAR_KYC' with the created parameter list
    }

    public PolicyHolderSearchResponse searchPolicyHolder(PolicyHolderSearchRequest request) {
        PolicyHolder policyHolder = policyHolderRepository.findPolicyHolderById(request.getPolicyHolderId());
        PolicyHolderSearchResponse response = new PolicyHolderSearchResponse();
        response.setPolicyHolder(policyHolder);
        return response;
    }

    public List<LanguageDTO> fetchLanguages() {
        return policyHolderRepository.findAllLanguages();
    }

    public void updateCommunicationLanguage(int languageId) {
        policyHolderRepository.updateLanguage(languageId);
    }

    public AgeCalculationResponse calculateAge(AgeCalculationRequest request) {
        int age = AgeCalculationUtil.calculateAge(request.getDateOfBirth(), request.getInceptionDate());
        AgeCalculationResponse response = new AgeCalculationResponse();
        response.setAge(age);
        return response;
    }
}